Add missing type hints to `core/dbt/contracts/sql.py`

### Housekeeping

- [X] I am a maintainer of dbt-core

### Short description

We're putting some effort into improving our typing. There are two type errors in `core/dbt/contracts/sql.py`, we should fix them.

```
core/dbt/contracts/sql.py:32: error: Function is missing a return type annotation  [no-untyped-def]
core/dbt/contracts/sql.py:32: note: Use "-> None" if function does not return a value
core/dbt/contracts/sql.py:43: error: Function is missing a return type annotation  [no-untyped-def]
```

### Acceptance criteria

The type issues are fixed

### Suggested Tests

N/A

### Impact to Other Teams

None

### Will backports be required?

No

### Context

_No response_

--- a/.changes/unreleased/Under the Hood-20240827-113123.yaml
+++ b/.changes/unreleased/Under the Hood-20240827-113123.yaml
@@ -1,0 +1,6 @@
+kind: Under the Hood
+body: Fix typing issues in core/dbt/contracts/sql.py
+time: 2024-08-27T11:31:23.749912-05:00
+custom:
+  Author: QMalcolm
+  Issue: "10614"

--- a/core/dbt/contracts/sql.py
+++ b/core/dbt/contracts/sql.py
@@ -29,7 +29,8 @@ class RemoteCompileResult(RemoteCompileResultMixin):
     generated_at: datetime = field(default_factory=datetime.utcnow)
 
     @property
-    def error(self):
+    def error(self) -> None:
+        # TODO: Can we delete this? It's never set anywhere else and never accessed
         return None
 
 
@@ -40,7 +41,7 @@ class RemoteExecutionResult(ExecutionResult):
     args: Dict[str, Any] = field(default_factory=dict)
     generated_at: datetime = field(default_factory=datetime.utcnow)
 
-    def write(self, path: str):
+    def write(self, path: str) -> None:
         writable = RunResultsArtifact.from_execution_results(
             generated_at=self.generated_at,
             results=self.results,

